src/rdkafka_telemetry_encode.c

```diff
@@ -0,0 +1,869 @@
+calculate_throttle_max(rd_kafka_t *rk, rd_kafka_broker_t *broker) {
+        rd_kafka_telemetry_metric_value_t max_throttle;
+        rd_kafka_broker_t *rkb;
+
+        max_throttle.intValue = 0;
+        TAILQ_FOREACH(rkb, &rk->rk_brokers, rkb_link) {
+                max_throttle.intValue =
+                    RD_MAX(max_throttle.intValue,
+                           rkb->rkb_avg_throttle.ra_v.maxv_interval);
+        }
+        max_throttle.intValue /= RDKAFKA_TELEMETRY_NS_TO_MS_FACTOR;
+        return max_throttle;
+}
+
+static rd_kafka_telemetry_metric_value_t
+calculate_queue_time_avg(rd_kafka_t *rk, rd_kafka_broker_t *broker) {
+        rd_kafka_telemetry_metric_value_t avg_queue_time;
+        int64_t sum_value = 0, broker_count = rk->rk_broker_cnt.val;
+        rd_kafka_broker_t *rkb;
+
+        TAILQ_FOREACH(rkb, &rk->rk_brokers, rkb_link) {
+                int64_t current_cnt = rkb->rkb_avg_outbuf_latency.ra_v.cnt;
+                int64_t historic_cnt =
+                    rkb->rkb_c_historic.rkb_avg_outbuf_latency.ra_v.cnt;
+
+                if (current_cnt > historic_cnt) {
+                        int64_t current_sum =
+                            rkb->rkb_avg_outbuf_latency.ra_v.sum;
+                        int64_t historic_sum =
+                            rkb->rkb_c_historic.rkb_avg_outbuf_latency.ra_v.sum;
+                        int64_t cnt_diff = current_cnt - historic_cnt;
+                        int64_t sum_diff = current_sum - historic_sum;
+
+                        sum_value +=
+                            sum_diff /
+                            (cnt_diff * RDKAFKA_TELEMETRY_NS_TO_MS_FACTOR);
+                }
+        }
+        avg_queue_time.intValue = sum_value / broker_count;
+        return avg_queue_time;
+}
+
+static rd_kafka_telemetry_metric_value_t
+calculate_queue_time_max(rd_kafka_t *rk, rd_kafka_broker_t *broker) {
+        rd_kafka_telemetry_metric_value_t max_queue_time;
+        rd_kafka_broker_t *rkb;
+
+        max_queue_time.intValue = 0;
+        TAILQ_FOREACH(rkb, &rk->rk_brokers, rkb_link) {
+                max_queue_time.intValue =
+                    RD_MAX(max_queue_time.intValue,
+                           rkb->rkb_avg_outbuf_latency.ra_v.maxv_interval);
+        }
+        max_queue_time.intValue /= RDKAFKA_TELEMETRY_NS_TO_MS_FACTOR;
+        return max_queue_time;
+}
+
+static rd_kafka_telemetry_metric_value_t
+calculate_consumer_assigned_partitions(rd_kafka_t *rk,
+                                       rd_kafka_broker_t *broker) {
+        rd_kafka_telemetry_metric_value_t assigned_partitions;
+        rd_kafka_broker_t *rkb;
+        int32_t total_assigned_partitions = 0;
+
+        TAILQ_FOREACH(rkb, &rk->rk_brokers, rkb_link) {
+                total_assigned_partitions += rkb->rkb_toppar_cnt;
+                total_assigned_partitions -=
+                    rkb->rkb_c_historic.assigned_partitions;
+        }
+        assigned_partitions.intValue = total_assigned_partitions;
```

Comment:
Given the metric name is `consumer.coordinator.assigned.partitions` we can use this, that is also used for the cgrp part of the stats cb.
```suggestion
        assigned_partitions.int_value = rk->rk_cgrp ?
            rk->rk_cgrp->rkcg_c.assignment_size
            : 0;
```

Reply:
Updated